### PR TITLE
[Tests-Only] [full-ci] add tests for trashbin restore to different location

### LINKED ISSUE #38882: add trashbin restore scenarios for when the original folder has been deleted
See comment https://github.com/cs3org/reva/pull/1795#issuecomment-866809751

We want to test what happens when a whole folder tree has been deleted, then the trashbin restore API is used to restore just 1 file or a whole sub-folder from within the deleted folder tree, and without specifying the location to restore it to.

Does the oC10 server automatically create the previous top-level folders needed? Or the item is restored to the root of the user's file system? Or an error status is returned because the target folder tree for the restore does not exist? Or?

### PR DESCRIPTION
**Pull request by dpakach**
## Description

Add tests for different cases of trashbin restore where the original location doesn't exists.

## Related Issue

- Fixes https://github.com/owncloud/core/issues/38882

## How Has This Been Tested?

🤖 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiWebdavMove1-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30822/96/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiTrashbin-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30826/88/1

File: tests/acceptance/features/apiTrashbin/trashbinRestore.feature [hunk 1]
```diff
@@ -440,3 +440,98 @@ Feature: Restore deleted files/folders
...
+    And user "Alice" should see the following elements
+      | /parent_folder/               |
+      | /parent_folder/sub/           |
+      | /parent_folder/sub/parent.txt |
+    Examples:
+      | dav-path |
+      | old      |
+      | new      |
+
+  @skipOnOCV10 @issue-398898
```
Comment (phil-davis) [inline tests/acceptance/features/apiTrashbin/trashbinRestore.feature:505]:
```suggestion
  @skipOnOcV10 @issue-398898
```

And issue number is too long.

File: tests/acceptance/features/apiTrashbin/trashbinRestore.feature [hunk 2]
```diff
@@ -440,3 +440,98 @@ Feature: Restore deleted files/folders
...
+    And user "Alice" should not see the following elements
+      | /parent_folder/               |
+      | /parent_folder/sub/           |
+      | /parent_folder/sub/parent.txt |
+    Examples:
+      | dav-path |
+      | old      |
+      | new      |
+
+  @skipOnOCV10 @issue-398898
```
Comment (phil-davis) [inline tests/acceptance/features/apiTrashbin/trashbinRestore.feature:524]:
```suggestion
  @skipOnOcV10 @issue-398898
```

And issue number is too long.

File: tests/acceptance/features/bootstrap/TrashbinContext.php [hunk 3]
```diff
@@ -584,6 +584,7 @@ private function restoreElement($user, $originalPath, $destinationPath = null, $
 			$destinationPath = $originalPath;
 		}
 		foreach ($listing as $entry) {
+			var_dump($entry['original-location']);
```
Comment (Talank) [inline tests/acceptance/features/bootstrap/TrashbinContext.php:587]: debug line?

File: tests/acceptance/features/apiTrashbin/trashbinRestore.feature [hunk 4]
```diff
@@ -440,3 +440,98 @@ Feature: Restore deleted files/folders
...
+    And user "Alice" should see the following elements
+      | /parent_folder/               |
+      | /parent_folder/sub/           |
+      | /parent_folder/sub/parent.txt |
+    Examples:
+      | dav-path |
+      | old      |
+      | new      |
+
+  @skipOnOcV10 @issue-398898
```
Comment (Talank) [inline tests/acceptance/features/apiTrashbin/trashbinRestore.feature:505]: Should the issue be `38898` or `398898`? I can not find any issue with issue number `398898`

File: tests/acceptance/features/apiTrashbin/trashbinRestore.feature [hunk 5]
```diff
@@ -440,3 +440,98 @@ Feature: Restore deleted files/folders
...
+    And user "Alice" should not see the following elements
+      | /parent_folder/               |
+      | /parent_folder/sub/           |
+      | /parent_folder/sub/parent.txt |
+    Examples:
+      | dav-path |
+      | old      |
+      | new      |
+
+  @skipOnOcV10 @issue-398898
```
Comment (Talank) [inline tests/acceptance/features/apiTrashbin/trashbinRestore.feature:524]: same here, I think the issue number is wrong.

File: tests/acceptance/features/apiTrashbin/trashbinRestore.feature [hunk 6]
```diff
@@ -131,8 +131,8 @@ Feature: Restore deleted files/folders
       | new      |
 
   @local_storage
-  @skipOnEncryptionType:user-keys @encryption-issue-42
-  @skip_on_objectstore
+    @skipOnEncryptionType:user-keys @encryption-issue-42
+    @skip_on_objectstore
```
Comment (phil-davis) [inline tests/acceptance/features/apiTrashbin/trashbinRestore.feature:135]: There is an indent change here. If you have to edit again anyway, then remove these spaces.

File: tests/acceptance/features/apiTrashbin/trashbinRestore.feature [hunk 7]
```diff
@@ -440,3 +440,98 @@ Feature: Restore deleted files/folders
...
+      | ETag | /^"[a-f0-9:\.]{1,32}"$/ |
+    And as "Alice" the file with original path "/parent_folder/sub/parent.txt" should not exist in the trashbin
+    And the content of file "parent.txt" for user "Alice" should be "parent text"
+    Examples:
+      | dav-path |
+      | old      |
+      | new      |
+
+
+  Scenario Outline: A deleted file inside a nested folder cannot be restored to the original location if the location doesn't exists
```
Comment (phil-davis) [inline tests/acceptance/features/apiTrashbin/trashbinRestore.feature:463]:
```suggestion
  Scenario Outline: A deleted file inside a nested folder cannot be restored to the original location if the location doesn't exist
```

File: tests/acceptance/features/apiTrashbin/trashbinRestore.feature (hunk 4 above)
Comment (dpakach) [reply tests/acceptance/features/apiTrashbin/trashbinRestore.feature]: yes, good find

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiTrashbin-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30827/88/1

**Review by phil-davis — APPROVED** [review]
Nothing important - I there is no other reason to make a change, then this is OK

**Review by phil-davis — APPROVED** [review]

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiTrashbin-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30830/88/1

File: tests/acceptance/features/apiTrashbin/trashbinRestoreOc10Issue38898.feature [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    And the value of the item "/d:error/s:exception" in the response about user "Alice" should be "Sabre\DAV\Exception\Forbidden"
+    And as "Alice" the file with original path "/parent_folder/sub/parent.txt" should exist in the trashbin
+    And user "Alice" should not see the following elements
+      | /parent_folder/               |
+      | /parent_folder/sub/           |
+      | /parent_folder/sub/parent.txt |
+    Examples:
+      | dav-path | dav_error_message                                                     |
+      | old      | Requested uri () is out of base uri (/%base_path%/remote.php/webdav/) |
+      | new      | Requested uri () is out of base uri (/%base_path%/remote.php/dav/)    |
```
Comment (phil-davis) [inline tests/acceptance/features/apiTrashbin/trashbinRestoreOc10Issue38898.feature:31]: If `base_path`  is empty then there is no double `/` in the `dav_error_message`

File: tests/acceptance/features/apiTrashbin/trashbinRestoreOc10Issue38898.feature [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    Then the HTTP status code should be "403"
+    And the value of the item "/d:error/s:message" in the response about user "Alice" should be "<dav_error_message>"
+    And the value of the item "/d:error/s:exception" in the response about user "Alice" should be "Sabre\DAV\Exception\Forbidden"
+    And as "Alice" the file with original path "parent.txt" should exist in the trashbin
+    And user "Alice" should not see the following elements
+      | /parent.txt               |
+    Examples:
+      | dav-path | dav_error_message                                                     |
+      | old      | Requested uri () is out of base uri (/%base_path%/remote.php/webdav/) |
+      | new      | Requested uri () is out of base uri (/%base_path%/remote.php/dav/)    |
```
Comment (phil-davis) [inline tests/acceptance/features/apiTrashbin/trashbinRestoreOc10Issue38898.feature:49]: If `base_path`  is empty then there is no double `/` in the `dav_error_message`

**Review by phil-davis — COMMENTED** [review]
Needs some adjustment for when `base_path` is empty.

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUIComments-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30842/129/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUIRestrictSharing-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30847/140/1

File: tests/acceptance/features/bootstrap/TrashbinContext.php [hunk 10]
```diff
@@ -606,6 +606,52 @@ private function restoreElement($user, $originalPath, $destinationPath = null, $
 		return null;
 	}
 
+	/**
+	 * @When user :user restores the folder with original path :originalPath without specifing the destination using the trashbin API
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/TrashbinContext.php:610]:
```suggestion
	 * @When user :user restores the folder with original path :originalPath without specifying the destination using the trashbin API
```

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUIRestrictSharing-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30870/140/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiTrashbin-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30887/88/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUIProxySmoke-3-1-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/30892/167/1

**Pull request merged**